cirq-core/cirq/ops/controlled_operation.py

```diff
@@ -134,7 +141,8 @@ def _extend_matrix(self, sub_matrix: np.ndarray) -> np.ndarray:
         sub_n = len(qid_shape) - len(self.controls)
         tensor = qis.eye_tensor(qid_shape, dtype=sub_matrix.dtype)
         sub_tensor = sub_matrix.reshape(qid_shape[len(self.controls) :] * 2)
-        for control_vals in itertools.product(*self.control_values):
+        for control_vals in self.control_values:
+            control_vals = cv.ControlValues.flatten(control_vals)
```

Comment:
the `|00><00| + |11><11|` combination can be added in the same way that the xor combination `|01><01| + |10><10|` which was not supported in the old way and can be seen in the tests

Reply:
Did we ever have a design review on this one? I provided the initial code reviews assuming a design was approved, but I still kind of stand by what I originally said on the linked issue, that this could be done more simply by just adding a bool flag to the ControlledGate constructor. That flag would control whether to interpret the 2D input array vertically as allowed states by qubit (which is how it currently works, but is limited), or horizontally as a full set of allowed states (which allows full control). 

That this question came up during code review makes me think end users are going to have trouble understanding how to use these classes.

Reply:
I introduced the idea in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-952284810 and talked about it ina cirq meeting https://github.com/quantumlib/Cirq/issues/4512#issuecomment-953208347, wrote a prototype in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-954142591 and got a review in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-974322956

--------------------------------------------------------------
the boolean idea has some issues
1. it doesn't address your previous comment  https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817, it fact it will make it harder to add those new ways lalter.
2. there is no way of getting around computing products, what a boolean will do is to force the users to compute them before passing them to the constructor, a work around would be passing an array of booleans but that has its own issues.
3. any work around will make the constructors for controlledGate and controlledOperation execture complex logic which is not a good idea in general, introduce tech debt and make it harder change or support new types later,

-----------------------------------------------------------------
lets discuss this in the next cirq meeting.

Reply:
Sounds good. I have limited availability for cynqs these days, but I think a reprise with the rest of the team on Wed would be great just to get the ball rolling on this PR, since it seems to keep going stagnant. I'm looking forward to the functionality.